intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/plugin/IdePluginManager.kt

```diff
@@ -0,0 +1,425 @@
+    descriptorPath: String,
+    validateDescriptor: Boolean,
+    resourceResolver: ResourceResolver,
+    parentPlugin: PluginCreator?
+  ): PluginCreator {
+    val libDir = File(root, "lib")
+    if (!libDir.isDirectory) {
+      return createInvalidPlugin(root, descriptorPath, PluginDescriptorIsNotFound(descriptorPath))
+    }
+    val files = libDir.listFiles()
+    if (files == null || files.isEmpty()) {
+      return createInvalidPlugin(root, descriptorPath, PluginLibDirectoryIsEmpty())
+    }
+    putMoreLikelyPluginJarsFirst(root, files)
+    val jarFiles = files.filter { it.isJar() }
+    val libResourceResolver: ResourceResolver = JarFilesResourceResolver(jarFiles)
+    val compositeResolver: ResourceResolver = CompositeResourceResolver(listOf(libResourceResolver, resourceResolver))
+    val results: MutableList<PluginCreator> = ArrayList()
+    for (file in files) {
+      val innerCreator: PluginCreator = if (file.isJar() || file.isZip()) {
+        //Use the composite resource resolver, which can resolve resources in lib's jar files.
+        loadPluginInfoFromJarFile(file, descriptorPath, validateDescriptor, compositeResolver, parentPlugin)
+      } else if (file.isDirectory) {
+        //Use the common resource resolver, which is unaware of lib's jar files.
+        loadPluginInfoFromDirectory(file, descriptorPath, validateDescriptor, resourceResolver, parentPlugin)
+      } else {
+        continue
+      }
+      results.add(innerCreator)
+    }
+    val possibleResults = results.stream()
+      .filter { r: PluginCreator -> r.isSuccess || hasOnlyInvalidDescriptorErrors(r) }
+      .collect(Collectors.toList())
+    if (possibleResults.size > 1) {
+      val first = possibleResults[0]
+      val second = possibleResults[1]
+      val multipleDescriptorsProblem: PluginProblem = MultiplePluginDescriptors(
+        first.descriptorPath,
+        first.pluginFileName,
+        second.descriptorPath,
+        second.pluginFileName
+      )
+      return createInvalidPlugin(root, descriptorPath, multipleDescriptorsProblem)
+    }
+    return if (possibleResults.size == 1) {
+      possibleResults[0]
+    } else createInvalidPlugin(root, descriptorPath, PluginDescriptorIsNotFound(descriptorPath))
+  }
+
+  private fun loadPluginInfoFromJarOrDirectory(
+    pluginFile: File,
+    descriptorPath: String,
+    validateDescriptor: Boolean,
+    resourceResolver: ResourceResolver,
+    parentPlugin: PluginCreator?
+  ): PluginCreator {
+    val systemIndependentDescriptorPath = descriptorPath.toSystemIndependentName()
+    return when {
+      pluginFile.isDirectory -> {
+        loadPluginInfoFromDirectory(pluginFile, systemIndependentDescriptorPath, validateDescriptor, resourceResolver, parentPlugin)
+      }
+      pluginFile.isJar() -> {
+        loadPluginInfoFromJarFile(pluginFile, systemIndependentDescriptorPath, validateDescriptor, resourceResolver, parentPlugin)
+      }
+      else -> {
+        throw IllegalArgumentException()
```

Comment:
else -> error() =)

Reply:
That will change exception type, I don't think it's a good idea